sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala

```diff
@@ -279,22 +275,40 @@ object OrcUtils extends Logging {
   }
 
   /**
-   * Given a `StructType` object, this methods converts it to corresponding string representation
-   * in ORC.
+   * Given two `StructType` object, this methods converts it to corresponding string representation
+   * in ORC. The second `StructType` used to change the `TimestampNTZType` as LongType in result
+   * schema string when reading `TimestampNTZ` as `TimestampLTZ`.
    */
-  def orcTypeDescriptionString(dt: DataType): String = dt match {
-    case s: StructType =>
+  def getOrcSchemaString(
+      dt: DataType, orcDt: Option[DataType] = None): String = (dt, orcDt) match {
+    case (s1: StructType, Some(s2: StructType)) =>
+      val fieldTypes = s1.fields.map { f =>
+        val idx = s2.fieldNames.indexWhere(caseSensitiveResolution(_, f.name))
```

Comment:
Is this a potential O(n^2) issue? If I am reading correctly, it says "for each field in s1, scan s2 to find the matching field".

Probably not a noticeable issue under most circumstances. But for wide structs with many input files, it could add up.

Reply:
`....map { f => ....indexWhere }` is widely used in Spark. I think it could simplify the code.

Reply:
>I think it could simplify the code

True, but it can be impactful.

Here are similar situations that bypass the O(n^2) complexity:

- `ParquetReadSupport #clipParquetGroupFields` [here](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetReadSupport.scala#L330) and [here](https://github.com/apache/spark/blob/f4543a545cbc4c99e210302858128f30f95011a9/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetReadSupport.scala#L340)
- `OrcUtils#requestedColumnIds` [here](https://github.com/apache/spark/blob/c5d7c3c17f140bd81ee886893d2e626644ebb6ac/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala#L253)

In a contrived query (6000 bigint columns, 65 files, 2000 rows per file, single executor) I saw a 40-50% penalty between the PR and the PR's base (08dd010860).

Reply:
OK. Updated.

Reply:
Thanks. That fixed the perf issue.